File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/log/ConversationEvents.kt
package com.intatis.shared.log

import com.intatis.shared.security.PermissionDecision
import com.intatis.shared.security.RiskLevel
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.io.File
import java.time.Instant
import java.util.concurrent.atomic.AtomicInteger

interface ConversationEventSink {
    suspend fun append(eventType: String, payload: Map<String, Any?>)
}

class NullConversationEventSink : ConversationEventSink {
    override suspend fun append(eventType: String, payload: Map<String, Any?>) {}
}

object ConversationEventKinds {
    const val ToolCall = "tool_call"
    const val ToolResult = "tool_result"
    const val PermissionRequest = "permission_request"
    const val PermissionResolved = "permission_resolved"
    const val PermissionReview = "permission_review"
    const val PatchProposed = "patch_proposed"
    const val AgentToAgentMessage = "agent_to_agent_message"
}

object ConversationEventPayloads {
    fun toolCall(toolCallId: String, agent: String, tool: String, args: String) = mapOf(
        "tool_call_id" to toolCallId,
        "agent" to agent,
        "tool" to tool,
        "args" to args,
    )

    fun toolResult(toolCallId: String, observation: String, truncated: Boolean? = null): Map<String, Any?> {
        val payload = mutableMapOf<String, Any?>("tool_call_id" to toolCallId, "observation" to observation)
        if (truncated != null) payload["truncated"] = truncated
        return payload
    }

    fun permissionRequest(
        requestId: String,
        agent: String,
        tool: String,
        args: String,
        risk: RiskLevel,
        reason: String,
    ): Map<String, Any?> = mapOf(
        "request_id" to requestId,
        "agent" to agent,
        "tool" to tool,
        "args" to args,
        "risk" to risk.name.lowercase(),
        "reason" to reason,
    )

    fun permissionResolved(
        requestId: String?,
        tool: String,
        agent: String?,
        decision: PermissionDecision,
        risk: RiskLevel,
        reason: String,
    ): Map<String, Any?> = mutableMapOf<String, Any?>(
        "request_id" to (requestId ?: ""),
        "tool" to tool,
        "agent" to (agent ?: ""),
        "decision" to when (decision) {
            PermissionDecision.ALLOW -> "allow"
            PermissionDecision.DENY -> "deny"
            PermissionDecision.ASK_USER -> "ask_user"
        },
        "risk" to risk.name.lowercase(),
        "reason" to reason,
    )

    fun permissionReview(
        tool: String,
        decision: PermissionDecision,
        risk: RiskLevel,
        reason: String,
        reviewerModel: String,
        agent: String? = null,
    ): Map<String, Any?> = mutableMapOf<String, Any?>(
        "tool" to tool,
        "agent" to (agent ?: ""),
        "reviewer_model" to reviewerModel,
        "decision" to when (decision) {
            PermissionDecision.ALLOW -> "allow"
            PermissionDecision.DENY -> "deny"
            PermissionDecision.ASK_USER -> "ask_user"
        },
        "risk" to risk.name.lowercase(),
        "reason" to reason,
    )

    fun patchProposed(patchId: String, agent: String, files: List<String>, diff: String): Map<String, Any?> = mapOf(
        "patch_id" to patchId,
        "agent" to agent,
        "files" to files,
        "diff" to diff,
    )

    fun agentToAgentMessage(from: String, to: String, content: String, mediated: Boolean): Map<String, Any?> = mapOf(
        "from" to from,
        "to" to to,
        "content" to content,
        "mediated" to mediated,
    )
}

class SessionEventLog(category: String, session: String? = null) : ConversationEventSink {
    private val sessionId = session ?: "${System.currentTimeMillis()}-${java.util.UUID.randomUUID()}"
    private val path = run {
        val root = File(System.getProperty("user.home"), ".intatis/logs/$category")
        root.mkdirs()
        File(root, "$sessionId.jsonl").absolutePath
    }
    private val mutex = Mutex()
    private val seq = AtomicInteger(0)
    private val json = Json { prettyPrint = false }

    fun path(): String = path

    override suspend fun append(eventType: String, payload: Map<String, Any?>) {
        val wrapper = toJsonObject(
            mapOf(
                "type" to eventType,
                "session" to sessionId,
                "seq" to seq.getAndIncrement(),
                "ts" to Instant.now().toString(),
                "payload" to payload,
            )
        )
        mutex.withLock {
            File(path).appendText(json.encodeToString(wrapper) + "\n")
        }
    }

    private fun toJsonObject(value: Map<String, Any?>): JsonElement =
        buildJsonObject {
            value.forEach { (k, v) -> put(k, toJsonElement(v)) }
        }

    private fun toJsonElement(value: Any?): JsonElement = when (value) {
        null -> JsonPrimitive("null")
        is String -> JsonPrimitive(value)
        is Number -> JsonPrimitive(value)
        is Boolean -> JsonPrimitive(value)
        is Iterable<*> -> buildJsonArray { value.forEach { add(toJsonElement(it)) } }
        is Map<*, *> -> buildJsonObject { value.forEach { (k, v) -> put(k.toString(), toJsonElement(v)) } }
        else -> JsonPrimitive(value.toString())
    }
}

File: Intatis-Windows/src/Intatis.Shared/ConversationEvents.cs
using System.Text.Json;

namespace Intatis.Windows.Shared;

public interface IConversationEventSink
{
    Task AppendAsync(string eventType, Dictionary<string, object?> payload);
}

public sealed class NullConversationEventSink : IConversationEventSink
{
    public Task AppendAsync(string eventType, Dictionary<string, object?> payload)
    {
        return Task.CompletedTask;
    }
}

public static class ConversationEventKinds
{
    public const string ToolCall = "tool_call";
    public const string ToolResult = "tool_result";
    public const string PermissionRequest = "permission_request";
    public const string PermissionResolved = "permission_resolved";
    public const string PermissionReview = "permission_review";
    public const string PatchProposed = "patch_proposed";
    public const string AgentToAgentMessage = "agent_to_agent_message";
}

public static class ConversationEventPayloads
{
    public static Dictionary<string, object?> ToolCall(string toolCallId, string agent, string tool, string args)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["tool_call_id"] = toolCallId,
            ["agent"] = agent,
            ["tool"] = tool,
            ["args"] = args,
        };
    }

    public static Dictionary<string, object?> ToolResult(string toolCallId, string observation, bool? truncated = null)
    {
        var payload = new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["tool_call_id"] = toolCallId,
            ["observation"] = observation,
        };

        if (truncated.HasValue)
            payload["truncated"] = truncated.Value;

        return payload;
    }

    public static Dictionary<string, object?> PermissionRequest(
        string requestId,
        string? agent,
        string tool,
        string args,
        RiskLevel risk,
        string reason)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["request_id"] = requestId,
            ["agent"] = agent,
            ["tool"] = tool,
            ["args"] = args,
            ["risk"] = risk.ToString().ToLowerInvariant(),
            ["reason"] = reason,
        };
    }

    public static Dictionary<string, object?> PermissionResolved(
        string? requestId,
        string tool,
        PermissionDecision decision,
        RiskLevel risk,
        string reason,
        string? agent = null)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["request_id"] = requestId,
            ["tool"] = tool,
            ["agent"] = agent,
            ["decision"] = decision switch
            {
                PermissionDecision.Allow => "allow",
                PermissionDecision.Deny => "deny",
                PermissionDecision.AskUser => "ask_user",
                _ => "ask_user",
            },
            ["risk"] = risk.ToString().ToLowerInvariant(),
            ["reason"] = reason,
        };
    }

    public static Dictionary<string, object?> PermissionReview(
        string tool,
        PermissionDecision decision,
        RiskLevel risk,
        string reason,
        string reviewerModel,
        string? agent = null)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["agent"] = agent,
            ["tool"] = tool,
            ["reviewer_model"] = reviewerModel,
            ["decision"] = decision switch
            {
                PermissionDecision.Allow => "allow",
                PermissionDecision.Deny => "deny",
                PermissionDecision.AskUser => "ask_user",
                _ => "ask_user",
            },
            ["risk"] = risk.ToString().ToLowerInvariant(),
            ["reason"] = reason,
        };
    }

    public static Dictionary<string, object?> PatchProposed(string patchId, string agent, IReadOnlyList<string> files, string diff)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["patch_id"] = patchId,
            ["agent"] = agent,
            ["files"] = files,
            ["diff"] = diff,
        };
    }

    public static Dictionary<string, object?> AgentToAgentMessage(string from, string to, string content, bool mediated)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["from"] = from,
            ["to"] = to,
            ["content"] = content,
            ["mediated"] = mediated,
        };
    }
}

public sealed class SessionEventLog : IConversationEventSink
{
    private readonly string _session;
    private readonly string _path;
    private readonly SemaphoreSlim _mutex = new(1, 1);
    private int _seq;

    public SessionEventLog(string category, string? session = null)
    {
        var logs = Path.Combine(ConfigStore.ConfigFolder, "logs", category);
        Directory.CreateDirectory(logs);
        _session = session ?? ($"{DateTime.UtcNow:yyyyMMddTHHmmssfff}-{Guid.NewGuid():N}");
        _path = Path.Combine(logs, $"{_session}.jsonl");
    }

    public string Path => _path;

    public async Task AppendAsync(string eventType, Dictionary<string, object?> payload)
    {
        var wrapper = new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["type"] = eventType,
            ["session"] = _session,
            ["seq"] = Interlocked.Increment(ref _seq) - 1,
            ["ts"] = DateTime.UtcNow.ToString("o"),
            ["payload"] = payload,
        };

        var line = JsonSerializer.Serialize(wrapper);
        await _mutex.WaitAsync();
        try
        {
            await File.AppendAllTextAsync(_path, line + Environment.NewLine);
        }
        finally
        {
            _mutex.Release();
        }
    }
}
